edgekv/src/file_ops.rs

```diff
@@ -0,0 +1,402 @@
+use chrono::Utc;
+use fs_extra::dir::DirOptions;
+use std::collections::BTreeMap;
+use std::fs::{File, OpenOptions, Metadata};
+use std::io::{BufReader, BufWriter, Seek, SeekFrom, Write, Cursor, Error, Read};
+use std::path::{Path, PathBuf};
+use crate::Result;
+use crate::datastore::{KeyDirEntry, KeysDir};
+use crate::errors::EdgeKVError;
+use crate::schema::{DataEntry, Decoder, Encoder, HintEntry};
+use fs2::FileExt;
+use std::sync::atomic::Ordering;
+use std::cmp::min;
+use bloomfilter::Bloom;
+
+const DATA_FILE_EXTENSION: &str = "data";
+const HINT_FILE_EXTENSION: &str = "hint";
+const BUFFER_FILE_EXTENSION: &str = "buff";
+const BLOOM_FILTER_FILE_EXTENSION: &str = "blmf";
+
+#[derive(Debug, Clone)]
+pub struct FilePair {
+    file_id: String,
+    data_file_path: PathBuf,
+    hint_file_path: PathBuf,
+    bloom_filter_file_path: PathBuf,
+}
+
+impl FilePair {
+    fn new(file_id: &str) -> Self {
+        Self {
+            file_id: file_id.to_string(),
+            data_file_path: Default::default(),
+            hint_file_path: Default::default(),
+            bloom_filter_file_path: Default::default(),
+        }
+    }
+
+    pub fn data_file_path(&self) -> String {
+        String::from(self.data_file_path.to_string_lossy())
+    }
+
+    pub fn hint_file_path(&self) -> String {
+        String::from(self.hint_file_path.to_string_lossy())
+    }
+
+    pub fn bloom_filter_file_path(&self) -> String {
+        String::from(self.bloom_filter_file_path.to_string_lossy())
+    }
+}
+#[derive(Debug)]
+pub struct Index {
+    file_id: String,
+    data_file_path: PathBuf,
+    hint_file_path: PathBuf
+}
+
+impl Index {
+    pub fn read(&self, entry_position: u64, size : usize) -> Result<DataEntry> {
+        let data = File::open(self.data_file_path.as_path())?;
+        let mut reader = BufReader::new(data);
+        reader.seek(SeekFrom::Start(entry_position));
+        let data_entry = DataEntry::decode(&mut reader)?;
+        if !data_entry.check_crc() {
+            return Err(EdgeKVError::CorruptData);
+        }
+        Ok(data_entry)
+    }
+
+    pub fn get_hints(&self) -> Result<Vec<HintEntry>> {
+        let hint = File::open(self.hint_file_path.as_path())?;
+        let mut rdr = BufReader::new(hint);
+        let mut hints = Vec::new();
+        while let Ok(hint_entry) = HintEntry::decode(&mut rdr) {
+            hints.push(hint_entry)
+        }
+        Ok(hints)
+    }
+
+    pub fn get_hint_entries(&self) -> Result<BTreeMap<Vec<u8>, KeyDirEntry>> {
+        let mut entries = BTreeMap::new();
+        let hint = File::open(self.hint_file_path.as_path())?;
+        let mut rdr = BufReader::new(hint);
+        while let Ok(hint_entry) = HintEntry::decode(&mut rdr) {
```

Comment:
Same here. Is this assuming that the only possible failure is because there is no more data left? aren't other errors possible? it should be more explicit.